### ISSUE TITLE
Development: Multi-Language Support

### ISSUE THREAD

**Issue by cord** [getgrav/grav#170]
This issue is set up to collect requirements and discuss ideas on how to support multi-language sites with grav.

**Comment by cord** [comment]
## Case 1: Multi-Domain Setup
Grav already supports a multi-domain setup where there is a domain per language with a single installation.

Using the Environment configuration http://learn.getgrav.org/advanced/environment-config you can set the home-directory by Server-Name, e.g.

user/www.mysite**.com**/config.system.yaml.

```
home:
  alias: '/home/en' 
```

user/www.mysite**.de**/config.system.yaml.

```
home:
  alias: '/home/de' 
```

and route to the markdown-files in the matching language.


### Pros

- already in place

### Cons

- requires a domain per language
- the content structure is not connected, you need to maintain the versions in parallel

**Comment by cord** [comment]
## Case 2: Single content structure (idea)

the content structure would be defined in the main language and translations would be added inside the structure
so starting with a single language structure like this:
```
/home
    /documentation
        product_a.md
        service_b.md
    /about
        team.md
        contact.md
        legal.md
```

the translations would be added within the same structure 

```
/home
    /documentation
        product_a.md
        de.produkt_a.md
        service_b.md
        de.service_b.md
        de.folder_name.md
    /about
        team.md
        de.team.md
        contact.md
        legal.md
        de.kontakt.md
        de.folder_name.md
```

the translation-files e.g. `de.produkt_a.md` could hold a reference to the original allowing to link between language versions of the same page

the de.folder_name.md files hold the translation of the folder, so followin fully translated URLs would be possible:

```
/about/contact
/ueber/kontakt
```

the main language could be used as fallback for missing translations, e.g.

```
/ueber/legal
```
would link to the en-version of page legal and show up in menu.


### Pros:

- single content structure
- easy to maintain as you don't need to switch between folder structures

### Cons:

- development required

**Comment by cord** [comment]
## Case 3: Single content structure with multiple language versions per file (idea)

the markdown syntax is enhanced to handle multiple language versions in one file

## Pros:

- single content structure, easy to maintain

## Cons:

- pages might get long and confusing as you would need to scroll between versions
- changes in one language is not easy to identify (git)

**Comment by Sommerregen** [comment]
Hi @cord,

as far as I can see cases 1 and 2 are looking good. Case 3 is a little bit problematic, since extending Markdown syntax will break up compatibility with other software tools.

I really like your your idea of how to structure the `user` directory into

- user/www.mysite.*com*/config.system.yaml
- user/www.mysite.*de*/config.system.yaml
- etc.

However, it would be good to have in one file. What do you think about altering the `config.system.yaml`?

```
home:
  language: "en"
  alias: "/home"               # Identical for all multi-language sites

# Depending on domain partially overwrites settings in "home:"
domains:
  www.mysite.com:
    language: "en"
  www.mysite.de:
    language: "de"
  etc.
```

Because then a fallback will be automatically included, provided the `user/pages` structure has the form

```
/home
    /documentation
        item.md
        item.de.md
        item.en.md
    /about
        team.md
        team.de.md
        contact.md
        contact.de.md
        legal.md
        legal.de.md
```

where the structure basically corresponds to canonical page structure, which is altered only in the `*.de.md` files like this

```
slug: path-to-my-alternated-content
language: "de"                # Optional since Grav knows anyway from *.de.md
                              # that the page is written in German
```

Then if **Grav** founds a multi-languaged page in the canonical folder structure, it can immediately deliver it. If not **Grav** will either load the page from `*.md` or `*en.md` depending on the user settings. Via page arguments in the URL like in the easiest case

```
http://grav.com/my-fantastic-page/
http://grav.com/my-fantastic-page/lang:en
http://grav.com/my-fantastic-page/lang:de
```

one can switch through all available languages.

### Pros:

- single content structure, easy to maintain
- nice fallback 

### Cons:

- A slug has to be set for each alternated content (once per `*.<langcode>.md` file)

**Comment by cord** [comment]
hi @Sommerregen,

looks all good to me!

I like the idea of adding e.g. lang:de to switch between language version, but assuming that

```
http://grav.com/my-fantastic-page/lang:de
```
would deliver the same content as
```
http://grav.com/meine-fantastische-seite/
```

would be duplicate content? or would it be possible to redirect to the unique language specific URL?

regarding cons:
> A slug has to be set for each alternated content (once per *.<langcode>.md file)

this would keep the content structure more easy to maintain as you immediately see the different language versions of a page. so more a pro for me ;)

**Comment by Sommerregen** [comment]
Hi @cord !

> I like the idea of adding e.g. lang:de to switch between language version, but assuming that
> ```
> http://grav.com/my-fantastic-page/lang:de
> ```
> would deliver the same content as
> ```
> would be duplicate content? or would it be possible to redirect to the unique language specific URL?```

Yeah you are right, it would be duplicate content. A redirection would be possible (if desired). For SEO considerations the best would also to set a canonical link or similar (I'm not used to such things), such that search engines won't decrease the score of your site.

> how to get forward? grav is new for me, so no idea where to start

You asked @rhukster . @mahagr will probably join. Depending on their plans they might help you providing some code initial code for you.

Unfortunately I never have worked together with them, but you always have the possibility to create pull requests. If you are feeling comfortable in digging into unfamiliar code, then browse through the Grav's source code. According to my considerations you have to modify at least the config class, the pages class and maybe the class, which is responsible for the routing. This can be done, though with average working time. If you have time, just join! :smile: :+1:

**Comment by drzraf** [comment]
just wanted to make you think about the "fallback-language" notion
(eg: case where contents are usually translated but some are not, a fallback-language (list) may be used)
It's more about generating list of contents (or choosing the alias to use in anchors), but may be worth considering.

**Comment by cord** [comment]
just realised the proposed content structure does not follow the **Grav** standard "Each page should be placed in its own folder" ( http://learn.getgrav.org/content/content-pages ), fixed this in the following specification draft

**Comment by cord** [comment]
# Specification Draft 1


## 1. Configuration

The configuration system allows to define languages and domain mappings to languages

```
home:
  language: "en"               # default language
  alias: "/home"               # Identical for all multi-language sites

# Depending on domain partially overwrites settings in "home:"
domains:
  www.mysite.com:
    language: "en"
  www.mysite.de:
    language: "de"
  etc.
```


## 2. Content Structure

Based on a content structure in the main language additional language versions can be added into the structure like following:


```
/home
    /documentation
        default.md
        default.de.md
    /about
        default.md
        default.de.md
```

the `default.<language>.md` files contain the translated page content and additional meta-data to define a language specific slug and the language (optional):

```
slug: path-to-my-alternated-content
language: "de"                # (not necessary*) / Optional since Grav knows anyway from *.de.md
                              # that the page is written in German
```

(*in general the `.<language>` part of the file extension identifies the translation language)


**Example:**

the german version of `my-fantastic-page/default.md` ist stored as `my-fantastic-page/default.de.md` and contains beside the translated content following metadata to define the slug:

```
slug: meine-fantastische-seite
```


**Grav** knows from **.de** part in the filename the page is written in german language

## 3. Connection between language specific URLs

From the content structure **Grav** should be able to find the connected language specific URLs, but there should also be an easy way to link to a language version w/o knowing its translated slug. While authoring /translating the content the links can be written using the default language slugs while the real slug can be changed/translated later. Using relative slug links ( http://learn.getgrav.org/content/linking#slug-relative ) the directory name should always work as fallback (but currently doesn’t, see bug https://github.com/getgrav/grav/issues/173 )


**Example:**

the english page (**.com**) has following URL:

```
http://grav.com/my-fantastic-page/
```

by adding `lang:de` to the URL

```
http://grav.com/my-fantastic-page/lang:de
```

the page requests is redirect to the language specific url (**.de**):

```
http://grav.de/meine-fantastische-seite/
```


the page parser should resolve the "real" URL when generating the output for SEO reasons.


## 4. Questions / ideas:

- how to handle tags / categories? (global dictionary for translations?)
- The translated version should be loaded as an overload of the default language version. E.g. unless specified different the tags set in default version are taken for any language version

**Comment by Sommerregen** [comment]
Hi @cord,

that's a thorough and concise specification draft we have so far. Well done! On the question of how to handle tags/ categories: to use a global dictionary would be a good idea, since we need it anyway in core, if we want to translate the rest of the page (menu, blocks etc.), too. I don't know what the best approach will be. I'm only familiar with the global dictionary approach (and the way how it translates text) from Drupal. Do you have some ideas?

**Comment by cord** [comment]
> to use a global dictionary would be a good idea, since we need it anyway in core, if we want to translate the rest of the page (menu, blocks etc.)

**Menu**: in my understanding the Title of menu items are defined in the .md files as well - so should already be solved with the current approach?

**Block**: not sure?

assuming we would  globally change the content file extension `.md` to `.de.md`(an all `.de.md` pages exit), then all should be shown in the DE version, right?

For the dictionary of tags / categories: I think the dictionary should only be applied when generating the output?

**Comment by robwent** [comment]
The main things for multi-language in my eyes and other thoughts
- translations need to be related so you can link from one language to another and also output hreflang tags in the head for seo reasons.
- urls: each language could be prefixed by the language tag and the native language left as is, but some users might want all languages prefixed.
- for the end user: people don't want workarounds like joomfish. When it's possible to translate a page they want it in the same admin screen, either an editor for each language or a tab for each language to translate. When there is one page you don't want to be looking in 2 separate places for the translation.

From an info architecture point of view it makes sense to keep translations in the same folder with a different affix for the language. It also makes it easy to relate pages and keep everything together in the admin area (when it arrives :))

**Comment by drzraf** [comment]
Is the case:
distinct languages, distinct structures **but** some content connections is out of scope?
Eg: only japanese version talks about tofu (blocks, content, taxonomy), but the US version talks about hokey. The rest (news, poney, and contact) are common and are connected.

There are a bunch of use-cases when it comes to i18n. Don't you think that the specification should cover a couple of API / toolkit (Twig, MD syntax) in order to be simple but flexible?

1. How to connect i18n content (subset of how to connect contents)
  iso suffixed files are good for similarly structured version but this isn't always the case. Those which don"t fit will prefer to link using pages' header on a case by case, or maybe even better if they have a lot of content, they'd define their languages/directory-tree mappings usings patterns and YAML.


2. Templates: <http://twig.sensiolabs.org/doc/extensions/i18n.html>
  A huge + would be that template for a given language take priority. Eg: base.html.twig may be common for all version, but grabing partials/metadata.de.html.twig and partials/metadata.de.html.twig (if they exists) would be nice. Otherwise base.html.twig would contain a big ```switch/case(lang == ) {% include ... $lang.html.twig (if exists) %}```

3. Blocks: maybe a Grav extension to easily adapt Twig find() and a couple of variables like _requested-language_, _default-languages-preferences_, _page-language_, and a couple of others, could match most needs and let plugins implement one or more of the many possibilities of structure.

4. I don't see where would be allowed the path/ prefix for languages (site.com/de site.com/es). If it's covered by the "domain" configuration section, then it should be renamed to "url".

**Comment by mahagr** [comment]
As @cord said, the easiest way to do languages is to create a multi-site setup, which basically contains single language per domain or directory. That should already work out of the box except for simple translations system, which is needed to translate core Grav, plug-ins and themes. It may also require some extra configuration and plug-in to bind the pages together to allow navigation between languages. The major drawback of using this method is that it may be really hard to figure out relationships between the pages, as every language equals creating a new site.

The second idea (revised as the draft) should also work great. It already defines connections between the languages, but also allows you to have content which is specific for a single language only. It also makes it very easy to see which parts are translated and which are not as well as allowing translators easily to compare languages or pick up their base language for the translation. The main issue with this method is slugs and paths -- how do you find the file to modify is the URL can be very different to the real path of the file.

Third idea I would just forget; it's really not a good idea to have multiple languages in a single file. We might consider having a common file for tags and other common header data and then translations for every language in their own files -- including English.

The main issue I'm seeing with the current draft is the performance implications of having every language inside a single structure. As Grav by default travels through all the directories by looking modified files, it would mean easily 5 times more files to be browsed just because of there's no way to make directory iterator to skip files that aren't needed. So either we need to rethink how we want to detect the changes or use more aggressive caching and hope nothing was changed.

**Comment by mahagr** [comment]
I want to make a comment on the original idea 2, which wasn't compatible with the directory structure requirement of Grav. While I agree that keeping the current directory structure (as described in Draft) is much cleaner way to handle languages than the original idea 2 which combines all the pages and languages into a single directory, I would want to remind that the requirement of having a single page per directory is not something that could not be changed or overridden. 

While right now directory structure is tightly coupled to page structure, it doesn't need to be that way. In fact it is something we probably need to think about. We already have exceptions on the rules as plug-ins and templates may define their own pages. Maybe in the future you could be able to choose how the files are located in the filesystem...

**Comment by bovisp** [comment]
Hello All

This is great stuff. Just want to add a couple of quick comments (as I should really still be sleeping!)

Just a little bit of background on me. I'm a Canadian Federal Government employee working in the Department of the Environment (known as Environment Canada). In my division, I'm basically the only tech guy and do everything from graphic design and page layout to audio/video recording and production to web design and development/programming and CMS/LMS Administration/front and backend development.

Our main CMS is (gasp!) SharePoint 2007 and I run the whole show when it comes to Moodle 2.8 for our LMS. Both have blogging capabilities, but are utterly unsuited for our needs, thus the idea for a separate blogging platform. WordPress is a non-starter in the government.I had no idea that flat-file CMS's existed until about four months ago. I instantly went with [Kirby](http://ketkirby.com) because of its excellent multi language support, but when I discovered Grav, even without multilanguage support (which I was promised would come soon) and an admin panel, I was hooked. The main reason for the switch was the image cropping/resizing capabilities built in. And its speed of course.

In the government we are blind by the National Languages Act, which basically means that any web content **must** be in both languages.

Okay, enough of that. On to the options:

## Option 1
This is a non-starter for me as I am only allowed one domain. I don't have the luxury of having multiple domains.

## Option 2
This is the method, by and large, that Kirby uses and makes the most sense to me. I know that it would require development, but it would be very obvious to the translation team in my division (and to me!), which pages have been made multilingual and which haven't.

## Option 3 
I agree with many other posters that this is not a very elegant solution and would lead to overly long files.

## Other Requests/Comments/Questions
* For the Admin panel, there should be an easy way to switch between languages when writing the content. Via a pulldown menu or something similar.
* How is the dictionary going to be implemented. Are there going to be translations in core and the ability to add translation files for any plugin? Will it be easy for somebody to alter translation files from within the /user directory? For example, I would be using the French translation from core, but would like to have the ability to alter the core translation in certain cases as French Canadian is slightly different than French?
* There should be a clearly documented way for developers to implement multilanguage switchers on the front end. By this I mean that there should be documented code snippets on [learn.getgrav.org](http://learn.getgrav.org) that show you how to create a language select menu that you can put at the top of your pages (in, say, a menu bar). Other UI options for language switching should be included as well as needed. In Kirby, I had to find these on Gist or in the discussion forums and that's not acceptable to me.

As usual, you are all doing a great job and I very much look forward to the result!

**Comment by cord** [comment]
@drzraf 

>  How to connect i18n content (subset of how to connect contents)

think this is in important point, maybe we should talk about "locale" instead of "language"? e.g. a help-text might be different for "en-UK" and "en-US" lets say for currency, legal situations etc. So beside a language version there also might be a localised version, also see http://getkirby.com/docs/languages/setup how it is done in Kirby (commercial).

When there are pages only for a subset of locales I think it should be controllable through meta-data configuration ("show this page only for locale 'de-AT, de-CH'")

> Templates: http://twig.sensiolabs.org/doc/extensions/i18n.html

I think translation of terms / language elements are most relevant for applications but less important for content?

> I don't see where would be allowed the path/ prefix for languages (site.com/de site.com/es)
> (...)then it should be renamed to "url".

good point, see http://getkirby.com/docs/languages/setup


@mahagr 
> I want to make a comment on the original idea 2, which wasn't compatible with the directory structure requirement of Grav. 
> While right now directory structure is tightly coupled to page structure, it doesn't need to be that way.

Actually while going through the **Grav** documentation my feeling was, that using a file-name to select a template seems to be wired. Also it seems to me for intuitive to set a slug as meta-data than having the slug as a directory-name: I would use directory names to structure my content and use the slug to determine how the links (SEO) look like. 

But however am sure there are good reasons to do it like the current solution?



@bovisp
thanks for pointing to Kirby, looks good but beside the feature-set it requires to purchase a (cheap) license. (Myself am happy to spent some money, but rather for extensions than the core system).

Requirements for the admin panel probably require a dedicated thread?

**Comment by mahagr** [comment]
Regarding on multiple domains to implement multi-site, you really do not need to have domains. I have already created working solution to implement it in directory level. So you can have separate site in /en and another one in /fr.

I think we really need more than just locales, because of you can have site that has different content per region, eg. Americas may have different products than Europe. Still you could have the same language (English, French, Spanish) available in both of them, or at least slight variation of it.

@cord 

I agree on using filenames to select how to render the page is somewhat odd, but in multilanguage using the same method (language tag in the file) actually feels natural and having only translations for the single page in a folder seems to be the way to go.

There is no real technical reason why we chose the current solution except that page could also have assets and modular content bound to a single page. So the decision was mostly made because of the other assets that the page may have had would have been in different location to the page content itself. Which wouldn't have been issue either (on technical side).

But as I said, Grav doesn't need to be locked to a single directory structure; there is really no technical reason other than the current implementation "assumes it" to have slugs as directory name and template file as the filename. You can already override template inside the file...

**Comment by drzraf** [comment]
On Mon, Apr 20, 2015 at 01:35:27AM -0700, Matias Griese wrote:
> The main issue I'm seeing with the current draft is the performance
> implications of having every language inside a single structure. As
> Grav by default travels through all the directories by looking
> modified files [...]

One reason Grav is so nice is that it broke for good with the dogma that
said "if it's not SQL then it's slow".
And still there's huge margin for improvements.

Eg:
- stop iterating directories except if "cache/force-reload.lock" exists
- in installation instruction add a line about the inotify daemon
  (incron) rule to add which will touch "cache/force-reload.lock" when any
  modification happens under pages/
That's it.

I'm pretty sure at this point performances should not keep any good
design to popup.
[note: I'm not saying that _this specific_ file tree design is "good" not "not"]

**Comment by bovisp** [comment]
@mahagr:

> Regarding on multiple domains to implement multi-site, you really do not need to have domains. I have already created working solution to implement it in directory level. So you can have separate site in /en and another one in /fr.

Are you saying that it is possible to have two different sites supported by one install of Grav or am I mistaken? If that is the case, then I would support Options 1 or two of the multilanguage specification. Perhaps we could chat about this on Gitter sometime. I don't want to go off topic here regarding the focus of this thread.

Regards,
Paul B.

**Comment by mahagr** [comment]
Its possible to have many different sites supported by a single Grav install. It is possible to do that by using domains, subdomains, path -- or you could use user session to create unique instance of the site. You can also choose to share or not to share information between the sites and use overrides over the main site.

The catch is that you need to write the code to make the choice between different sites, but the good thing is that you're not only allowed to choose your own multi-site implementation, but you can also do all kinds of cool things, like have demo site with admin and do that in a way that every user has their own instance of the site to play with. Only your imagination has limits on this one. :)

**Comment by Sommerregen** [comment]
I'd like to add some thoughts, which hasn't been discussed yet or only mentioned once without considering them deeper.

I think, the specification draft 1 can been in terms of grav multi-site capabilities, unfortunately breaking any connections between pages which belong together. But can be used whenever a different page structure of each language is needed. Ok, connections can be restored, but then a different instance is needed to manage the connections and to keep them up-to-date when the page structure and contents are modified.

Locale prefixing seems to be good: the page structure tightly binds multiple languages of one page in e.g. one folder (but like @mahagr said, it doesn't need to be realized in such a way in the final draft/design decision).

@cord:
> **Menu:** in my understanding the Title of menu items are defined in the .md files as well - so should already be solved with the current approach?

Good point!

@drzraf
> A huge + would be that template for a given language take priority. Eg: base.html.twig may be common for all version, but grabing partials/metadata.de.html.twig and partials/metadata.de.html.twig (if they exists) would be nice. 

From the (plugin) developer point of view, I'm feeling incapable of providing a plugin template file for all languages. According to the DRY (don't repeat yourself) principle the template files for all languages share the same logic and thus are redundant. But a plugin has to be designed independent of a language, which has been later used.
This is different from the user point of view, who likes to customized the templates for its own needs and likes to adjust templates for the specific of the respective locales. Therefore, this would be definitely great to have such a feature!

Instead of maybe using http://twig.sensiolabs.org/doc/extensions/i18n.html with the gettext support (which seems really weird for me to implement and I'm not sure, whether everyone has this module installed) one could provide a TWIG function (maybe `t(message, params)`?) or emulate the TWIG block `trans`, which translates text to the current locale in the template?

@bovisp: Yeah, Kirby is great and probably the best competitor among Grav.

> How is the dictionary going to be implemented. Are there going to be translations in core and the ability to add translation files for any plugin? Will it be easy for somebody to alter translation files from within the /user directory?

The dictionary can be in the easiest case a list containing locales and translations from a (canonical) preferred language. The core could provide functions to access the locales and the respective translations e.g. via the mentioned `t(message, params)` function earlier. There should be, of course, the possibility to add/modify/remove translations and locales using probably `.po` files. Translations can thus be altered and most desirable from the user directory either by plugins or by the user itself.

**Comment by cord** [comment]
ok, a lot of ideas are exchanged - can we nail down a specification for multi-language support? Or is this thread already good enough?

**Comment by mahagr** [comment]
I think we could summarize the plan in the first message as I'm already confused of what we have agreed or not. Personally I would love to see both separate sites and translations in same directory approaches get implemented as both of them has their uses, even in a single site. We also need to solve issue where you may have 2 different sites for the same language (maybe using different locale), chosen by geolocation. 

Here is my use case (seen in many companies sites):

* main site in Finnish and Swedish (both official languages here)
* shorter site in English with main content, but maybe not blog and news
* one page sites for a few other languages (German, French, Russian...)

Regarding to translations, Grav source code already includes a class for reading gettext.

**Comment by cord** [comment]
@mahagr 

> Personally I would love to see both separate sites and translations in same directory approaches get implemented as both of them has their uses, even in a single site.

what about the idea to black/white list pages through meta-data for certain languages / locales ?

that should give best flexibility while keeping things in a common structure easy to maintain?

**Comment by mahagr** [comment]
Would it be simpler to just skip missing pages or if fallback language is defined, use that language instead. If fallback language doesn't have the page, you would just get 404.

I still believe that we need both multi-site and multi-language features together in order to implement all the features that may be needed.

Nevertheless, I don't think that they really contradict with each other, we just need to keep both use cases in the mind. And BTW, multi-site allows sites to share pages, but to have their own overrides as well.

**Comment by rhukster** [comment]
Some really great ideas here guys.  

I think the next step is for us (the core developers) to have a little chat about the implications of all the points raised.  We have to think what is the best approach that will work with Grav, while still providing the features and capabilities that are needed and have been requested.  All the while being maintainable and performant. We also have to think about how a multilang solution will impact the admin and other third-party plugins.

Over the next few weeks (in between everything else) we'll work on a draft document that goes into the detail we need in order to make this a reality. This will be more focused on how to 'code' the solution, but will also cover the capabilities, so I'm sure you will find it a useful read.

Keep your ideas and comments coming here, we will continue to monitor and interact with this thread.

Cheers,

**Comment by nickbe** [comment]
Just wanted to mention that I think it's important to have a domain dependend approach (de.site.com, en.site.com) as well as a simple manual in-site switch option. Also grav should try to detect the browser language and select an appropriate site language by itself if possible (or use a fallback language).

**Comment by cord** [comment]
Hey andy,

wonder if there is any progress for supporting multiple languages? If the content format is fixed, am happy to build the wordpress exporter as as WP plugin.

**Comment by rhukster** [comment]
No more progress yet because we are focusing on the admin plugin.  This 'thread' has proved immensely helpful though and we are sure to use the ideas there to form our ultimate solution.  Please keep adding to it if you have any more thoughts or ideas.

**Comment by rhukster** [comment]
Some good news! I'm pretty close to making multilang available for your testing.  i'll have to document all the features/capabilities and possible implications.  The good news is that this implementation does not add any performance degradation!  The actual solution is pretty close to what has been discussed here:

1) configure langs supported in system.yaml
2) default lang is first lang in array of supported langs
3) can explicitly reference lang with url, eg: http://yoursite.com/en/blog
4) each translation sits in the same directory (`default.en.md`, `default.fr.md`)
5) each page can have it's own slug (and combined with new page routes, have complete custom routing)
6) fallback to other supported languages if translation does not exist
7) can disable a page, but adding just `published: false` in md file
8) you are not restricted to just en, fr, etc, you can use locae style: `en-GB`, `en-US`, `fr-FR`, etc, just need to be consistent
9) supports optional lang specific twig files `en/default.html.twig` will be used before `default.html.twig` if found
10) translations will be stored in .yaml files inside configuration, eg: `config/languages/en.yaml`. So this will allow you to override anything in your own `user/config/languages` and also in plugins, themes, etc.

last bit i'm working on is the Twig filter and php function to allow these translations in twig/php. syntax will be: `{{ "DEFAULT_FOOTER_TEXT"|t }}` and you can use placeholders and variables too with sprintf type syntax: 

```
DEFAULT_FOOTER_TEXT: Grav is %s with %s, developed by <a href="http://rockettheme.com">RocketTheme.com</a>
```

```
{{ "DEFAULT_FOOTER_TEXT"|t('coded', 'love') }}
```

Of course, the string doesn't have to be capitals, that's just a mechanism to make them more visible in the code.

**Comment by cord** [comment]
sounds great! looking forward to see the documentation of the content structure - for august I plan to finalise the wordpress plugin to export qtranslate sites into grav multi language content structures for easy migration keeping all URLs in place.

**Comment by robwent** [comment]
How flexible is 6?
It's not always ideal to show a user content in another language if their language doesn't have a translation.
There could be a default message for each language saying the content isn't available, or just send them to a 404.

**Comment by rhukster** [comment]
you can simply put:
```
---
published: false
---
```

in a page for an unsupported language.. This will be the same as the page not being available at all.

**Comment by giansi** [comment]
This is something I really need. Can't wait to test it.

**Comment by bovisp** [comment]
Andy, I can't believe you came up with this so quickly. Your solution is exactly what I was hoping for. I would be more than happy to help with testing as I have experience with this workflow from my use of Kirby.

**Comment by rhukster** [comment]
Develop branch updated with latest.  Also this is the draft of the docs:

https://github.com/getgrav/grav-learn/blob/feature/0.9.30-updates/pages/02.content/09.multi-language/docs.md

**Comment by Sommerregen** [comment]
Hi @rhukster ,

haven't tried it yet, but I have a nagging question to you concerning [translation support](https://github.com/getgrav/grav-learn/blob/feature/0.9.30-updates/pages/02.content/09.multi-language/docs.md#translation-support). I really like the idea to put all strings into a separate translation folder into a YAML structure (at least into a separate fie, see below), but [translation via Twig](https://github.com/getgrav/grav-learn/blob/feature/0.9.30-updates/pages/02.content/09.multi-language/docs.md#translation-via-twig) really seems awkward for me.

Using capital strings as placeholders is not the standard way how translatng strings works (afaik). I think the better choice would be to use `pot` an `po` files (like i.e. Drupal), which are far better to maintain than simple YAML structures. Further twig files then become more readable. Instead of

```twig
{{ "SIMPLE_TEXT"|t(12, "London Zoo") }}
{# By the way in the docs there is a typo. A closing parenthesis is missing... and comments are enclosed like here with the dollar symbol and not the slashes. #}
```

we can write

```twig
{{ "There are %d monkeys in the %s."|t(12, "London Zoo") }}
```

which is concise, easy to read in the twig file (you don't have to figure out, what `"SIMPLE_TEXT"` means and needs for params etc.), you **don't exploit strings as placeholders** and this approach enables you to write a simple program extracting all strings to translate into a separate file. The last one is not really an argument for doing it, since you can do it with your approach, too. The same goes for the plugin support.

I really like to hear your opinion about that. Maybe this is not the best approach and one can do better. Thus, see it as a draft.

**Comment by cord** [comment]
Andy,
following comments/questions to your nice documentation draft:

**map language to domains**
is there a way to map the domain to a language, e.g. like this:
`http://yoursite.com` loads the `en` version
`http://yoursite.fr` loads the `fr` version

without using `/en/`or  `/fr/` in the url?

**linking between language versions**
would be great if each pages "knows" the url of other language versions to allow comfortable language switching on a page level or if just specifying the language would redirect, e.g.

`http://yoursite.fr/animals/mammals/bears`
redirects to
`http://yoursite.fr/animaux/mammiferes/ours`

**fallback for locales**
In my view locale versions (e.g. en-GB / en-US) would be used for pages with content different from the common "main" language version. So only these pages would get a localised version while all other common pages would be identically. To avoid duplicating the common pages into each locale there should be the definition of the fall back language, e.g. like this:

```
languages:
  supported:
    - en-GB
       fallback: en
    - en-US
       fallback: en      
    - en
    - fr
```
maybe the fallback could also just be the first part of the locale?

**Comment by rhukster** [comment]
@Sommerregen The main advantage over using YAML is 4 fold:

1) It's a consistent technology that we are already using and our uses will be familiar with
2) We have sophisticated compiling and caching mechanisms for YAML already that means its going to be fast (which as you know is key in Grav)
3) YAML files are human readable and don't require editors
4) Our YAML capabilities already have support for merging and overriding

I don't really see a compelling reason to use PO files, but if you have some good reasons, please let me know!

Regarding the use of uppercase constant-style key values, again this is based on my experience with Joomla, this is the system they use and frankly I like it :).  The problem with using 'english' or native language strings for keys is that if you want to change the native language key, you have to change the key in all the other languages.  This is a royal pain.

Maybe i'm misunderstanding how this would work with other languages ?

I think lowercase constant style is ok also, but i just like how the uppercase stands out in a Twig file, makes them easy to find.

**Comment by rhukster** [comment]
@cord, 

* **map language domains:** You can certainly do the domain redirect from `.htaccess`, It really can't be done via PHP alone.

* **link between language versions:** I have added a 'raw_route` attribute to page.. This is basically a route that is built from language + folders and will not change, no matter what route overrides you have.

so a page: `01.category/03.section/my-custom-page/default.en.md` and another with `default.fr.md` would always be accessible via: `/en/category/section/my-custom-page` or `/en/category/section/my-custom-page`.

I intend to write a simple language switcher plugin (today?) that you can use in your themes to that will know the current page, and be able to switch to another language version of the same page utilizing this new route.

* **fallback for locales** I like the idea for falling back to first part of locales, but will have to look at the implementation implications of doing that before I can 100% commit to doing it :)

**Comment by cord** [comment]
@rhukster 

> map language domains:
>You can certainly do the domain redirect from .htaccess, It really can't be done via PHP alone.

wonder if the configuration system could be enhanced like following to map languages to domains?

user/www.mysite**.com**/config.system.yaml.
```
home:
  language: 'en' 
```
user/www.mysite**.fr**/config.system.yaml.
```
home:
  language: 'fr' 
```

**Comment by rhukster** [comment]
With this system there is only **one** configuration file, in fact there is only one site, its not a multisite solution as that is actually much more complex to administer.  It's maintains its speed by caching each collection of pages, routes, and taxonomy **per-language**, and only as the language is requested.

I know it was discussed but really the intention of that was to provide different structure as well as content.  That can be achieved via this single site mechanism however.

You can still do a multi-site multi-lang solution using a mix of Grav's existing multisite streaming (needs to be documented still)  combined with this multilang solution, but frankly it's going to be quite complex, and not going to buy you anything.

**Comment by cord** [comment]
sorry, forgot to remove the `home`setting, was probably confusing. Am thinking of just setting the language using a domain configuration file like following:

user/www.mysite**.com**/config.system.yaml.
```
language: 'en' 
```
user/www.mysite**.fr**/config.system.yaml.
```
language: 'fr' 
```

so everything else would stay with the current approach (totally agree). I understand that giving only one configuration setting in the domain configuration file all would allow to overload only this single setting?

**Comment by rhukster** [comment]
Hmm... I see what your saying.. That might work as-is.. Not actually tested it :)

**Comment by bovisp** [comment]
First of all, I want to say **bravo** to you Andy and your team for making this happen so quickly!

This is a little off topic, but does concern translation-related issues.

In Kirby, they give you the option of attaching markdown files to media files (images, videos ect.). So with a folder structure like this complete with images:

```
- 01.Animals
  - default.en.md
  - default.fr.md
  - 01.Mammals
    - default.en.md
    - default.fr.md
    - 01.Bats
      - default.en.md
      - default.fr.md
      - bats.jpg
    - 02.Bears
      - default.md
      - bears.jpg
    - 03.Foxes
      - default.en.md
      - foxes.jpg
    - 04.Cats
      - default.en.md
      - default.fr.md
      - foxes.jpg
  - 02.Reptiles
  - 03.Birds
  - 04.Insets
  - 05.Aquatic
```
each image could also have its own markdown file with YAML frontmatter provided that the file contained the image name. So, for example, `bears.jpg` would have an associated `bears.jpg.md`.

The image files themselves could also have `.en` and `.fr` so that, for example, `bears.en.jpg` would only be shown in the English view and `bears.fr.jpg` would only be shown in the French view. The associated markdown files would thus look like `bears.en.jpg.md` and `bears.fr.jpg.md` for English and French respectively.

There would, of course be default fallbacks in the same way as you described in the Grav languages documentation. For example, if `en` is the current language, Grav would look for `bears.en.jpg` and the associated `bears.en.jpg.md` first. If that didn't exist, it would look for `bears.jpg` and the associated `bears.jpg.md`

Putting all of this together, we can modify the `\animals\mammals\bats\` folder structure to look something like:

```
- 01.Animals
  - 01.Mammals
    - 01.Bats
      - default.en.md
      - default.fr.md
      - bats.en.jpg
      - bats.fr.jpg
      - bats.en.jpg.md
      - bats.en.jpg.md
```
Now, contained within these media markdown files would be YAML frontmatter specific to that particular media file:

```
---
Title: A beautiful picture of bats
FigureNumber: 1
Author: Some Guy
Location: Texas
Year: 2015
---
```
This would beneficial for a portfolio gallery that uses a card-based design to display many images in a gallery page. The text in the image card would look something like this:

![](http://5to9design.ca/card.png)

where the text is being drawn from the YAML front matter drawn from that images markdown file.

Another benefit of this is that your SimpleSearch plugin could also be modified to look for words within those media markdown files. The search results page could then be used to display page results in one tab and and image (or media) results in another tab, very much like Google, Yahoo or Bing search results.

What's your opinion on this?

**Comment by rhukster** [comment]
I think that's a worthy addition.  I do think however, that it probably can wait till a later version as using the media metadata files at all is quite advanced, and not many people are even using them yet.

Would be awesome to get a user contributed PR for that work once the lang stuff has been finalized and released.

Regarding simplesearch, there is definitely room for a more plugable search engine system that could plugin into different places to search content, but I think that its best to keep simplesearch, simple :)

**Comment by bovisp** [comment]
Andy,

Glad that you like the idea. I can wait for this for sure. I would love to create this functionality in Grav, but I only know enough PHP to be dangerous (as my IT department says). We do have a PHP developer who is hacking Moodle for us currently, but, if I arm twist my manager, I might be able to convince him to give me the developer's time for this. No promises though.

Cheers,
Paul B.

**Comment by rhukster** [comment]
I would appreciate the help :)  After I get a few minor things finished I want to release 0.9.30 with this multilang and route stuff as it's a big improvement.  After that I'm going to jump back into the admin with both feet.

**Comment by Sommerregen** [comment]
@cord 

> **map language to domains**
> is there a way to map the domain to a language, e.g. like this:
> http://yoursite.com loads the en version
> http://yoursite.fr loads the fr version

This is doable with Grav right now, using  a little bit of the multi-site capabilites. Just name a file `setup.php` at the root, which will do the checking and so on. However it is a little bit complicated to do it and too long to explain here. Maybe this would be a good blog post, but at the moment my time is really limited to write one. @rhukster The same goes for a shortcodes tutorial :-/

@rhukster : I didn't want to bother you. It was just an idea, since I have good experiences with PO files. Maybe your approach can be as powerful as PO files, too. And you have striking arguments for using YAML structure. I never have used Joomla, so that's the reason why I'm feeling awkward concerning "string constants", but your point is again well-grounded. And yes also in other languages you have to change the keys then.

Thank you for letting me know the thoughs behind the design decision. As you know I'm still learning ;-)

**Comment by rhukster** [comment]
@cord BTW, can confirm that you can set a different language order for different environment config.yaml overrides and that works just fine :)

**Comment by rhukster** [comment]
With the release of Grav 0.9.30, **most** of these features have been implemented.  I'll close this for now, but feel free to create new bugs or requests as individual issues.
